File: src/main/java/com/conference/website/api/EngagementController.java
package com.conference.website.api;

import com.conference.website.dto.*;
import com.conference.website.service.EngagementService;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

@Validated
@RestController
@RequestMapping("/api/talks/stats")
public class EngagementController {

    private final EngagementService engagementService;

    public EngagementController(
                               EngagementService engagementService) {
        this.engagementService = engagementService;
    }

    @PostMapping("/{talkId}/engagement")
    public Mono<EngagementCountDto> recordEngagement(@PathVariable Long talkId, @RequestBody EngagementUpdateRequest request) {
        return engagementService.recordEngagement(talkId, request);
    }

    @GetMapping("/{talkId}/engagement")
    public Mono<EngagementCountDto> getEngagement(@PathVariable Long talkId) {
        return engagementService.getCurrentEngagement(talkId);
    }

}

File: src/main/kotlin/com/conference/website/api/EngagementController.kt
package com.conference.website.api

import com.conference.website.dto.*
import com.conference.website.service.EngagementService
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono

@Validated
@RestController
@RequestMapping("/api/talks/stats")
class EngagementController(
    private val engagementService: EngagementService
) {

    @PostMapping("/{talkId}/engagement")
    fun recordEngagement(
        @PathVariable talkId: Long,
        @RequestBody request: EngagementUpdateRequest
    ): Mono<EngagementCountDto> =
        engagementService.recordEngagement(talkId, request)

    @GetMapping("/{talkId}/engagement")
    fun getEngagement(@PathVariable talkId: Long): Mono<EngagementCountDto> =
        engagementService.getCurrentEngagement(talkId)
}
